### ISSUE TITLE
High convergence deltas with DeepLIFT/SHAP despite accounting for non-linearities.

### ISSUE THREAD

**Issue by Dflorizone** [jmschrei/tangermeme#47]
Hello, thanks for the great library!

I'm trying to implement deep_lift_shap on a deep residual network, but have struggled with high convergence deltas. I believe all the non-linear layers in the model (ReLU,Softmax) are registered by default, so I'm struggling to figure out why my convergence deltas are so high. Here is the model code:
```python
class ResidualBlock(nn.Module):
    def __init__(self, in_channels, out_channels,kernel_size,dilation=1, activation='relu',bn_momentum=0.01):
        super().__init__()
        self.in_channels, self.out_channels, self.activation = in_channels, out_channels, activation
        paddingAmount = int(dilation*(kernel_size-1)/2)
        self.convlayer1 = nn.Conv1d(in_channels=self.in_channels, out_channels=self.out_channels, kernel_size= kernel_size,dilation=dilation,stride=1,padding=paddingAmount,padding_mode='zeros')
        self.convlayer2 = nn.Conv1d(in_channels=self.in_channels, out_channels=self.out_channels, kernel_size= kernel_size,dilation=dilation,stride=1,padding=paddingAmount,padding_mode='zeros')
        self.bn1 = nn.BatchNorm1d(self.in_channels, momentum=bn_momentum)
        self.bn2 = nn.BatchNorm1d(self.in_channels, momentum=bn_momentum)
        self.shortcut = nn.Identity()
        self.activation1 = nn.ReLU()
        self.activation2 = nn.ReLU()   
    
    def forward(self, x):
        residual = x
        residual = self.shortcut(x)
        x = self.activation1(self.bn1(x))
        x = self.convlayer1(x)
        x = self.activation2(self.bn2(x))
        x = self.convlayer2(x)
        x = x + residual
        return x

class ResComboBlock(nn.Module):
    def __init__(self, in_channels, out_channels,res_W,res_dilation,bn_momentum=0.01):
        super().__init__()
        self.comboBlock = nn.Sequential(
        ResidualBlock(in_channels,out_channels,res_W,res_dilation,bn_momentum=bn_momentum),
        ResidualBlock(in_channels,out_channels,res_W,res_dilation,bn_momentum=bn_momentum),
        ResidualBlock(in_channels,out_channels,res_W,res_dilation,bn_momentum=bn_momentum),
        ResidualBlock(in_channels,out_channels,res_W,res_dilation,bn_momentum=bn_momentum)
        )
    
    def forward(self, x):
        return self.comboBlock(x)

class SpliceAI_10K(nn.Module):
    def __init__(self,CL_max,crop=True, **kwargs):
        super().__init__()
        n_channels = 32
        self.CL_max = CL_max
        self.crop = crop
        self.res_W = [11,11,21,41]
        res_dilation = [1,4,10,25]
        self.kernel_size = 1
        self.conv_layer_1 = nn.Conv1d(in_channels=4, out_channels=n_channels, kernel_size= self.kernel_size,stride=1)
        self.skip_layers = nn.ModuleList([nn.Conv1d(in_channels=n_channels, out_channels=n_channels, kernel_size= self.kernel_size,stride=1) for i in range(5)])
        self.res_layers = nn.ModuleList([ResComboBlock(in_channels=n_channels, out_channels=n_channels, res_W=self.res_W[i], res_dilation=res_dilation[i]) for i in range(4)])
        self.conv_final = nn.Conv1d(in_channels=n_channels, out_channels=3, kernel_size= self.kernel_size,stride=1)
        
    def forward(self, features):
        x = self.conv_layer_1(features)
        skip = self.skip_layers[0](x)

        for i,residualUnit in enumerate(self.res_layers):
            x = residualUnit(x)
            skip = skip + self.skip_layers[i+1](x)
        
        if self.crop:
            x = skip[:,:,self.CL_max//2:-self.CL_max//2]
        else:
            x = skip[:,:,:]
        x = self.conv_final(x)
        m = nn.Softmax(dim=1)
        return m(x)
```

The model takes an input tensor of shape (batch_size,4,15000) and outputs a tensor of shape (batch_size,3,15000). I wrapped the model in a class to produce a model prediction of shape (batch_size, n_targets). Here is the code I used:
```python
class WrappedModel(nn.Module):
    def __init__(self, base_model, target_idx):
        super().__init__()
        self.base_model = base_model
        self.target_idx = target_idx

    def forward(self, x):
        out = self.base_model(x)
        scalar = out[:, self.target_idx[0], self.target_idx[1]]
        return scalar.unsqueeze(1) # shape batch_size,n_targets

wrapped_model = WrappedModel(model, target_idx=(1, 4290))
X_attr = deep_lift_shap(model=wrapped_model,X,device='cpu',random_state=0,print_convergence_deltas=True)
# Where X is some one-hot encoded input data with shape 1,4,15000
```
This code produces convergence deltas around ~0.8. I noticed another similar issue with high convergence deltas #13, but it seems this was due to an element wise x^2 operation, which was the source of the non-linearity. It doesn't seem that the model I'm using has any other non-linear layers that are not being accounted for by default. I'm wondering if you have any insight into why my convergence deltas are so high. 

Thanks!

**Comment by Dflorizone** [comment]
@caenrigen Thanks for the suggestion! I tried replacing nn.ModuleList with the individual layers, but unfortunately still get the same convergence deltas. My model output is very close to 1, so there must be something else going wrong to produce such high convergence deltas relative to my output. I'll update if I solve the issue.

**Comment by jmschrei** [comment]
@Dflorizone sorry you're encountering this issue. Resolving these oddities an underappreciated challenge with running DeepLIFT, and most other packages (Captum, SHAP, etc) just silently fail. 

I think the issue is likely because you are creating a `nn.Softmax` within the forward function. All layers have to be defined in the `__init__` because they need to have hooks assigned to them before the DeepLIFT procedure begins.

If that doesn't work, I would recommend the same things that @caenrigen did. First, try to write out the entire model as a single torch.nn.Module object. Then, try removing non-linearities until the deltas are near-zero to isolate which one is the issue. I've found that the softmax operation can sometimes be problematic despite being registered -- I assume because you're squishing something with a very large range down to something where machine precision might play a bigger role? I usually don't include softmax operations in my model and, instead, output the raw logits and use a loss function that can accept logits. But sometimes it's fine. Once you've gotten *something* that works, even if it's not your main model, slowly add the complexity back in checking each time that the deltas are still not high.

**Comment by Dflorizone** [comment]
@jmschrei @caenrigen Thanks for the recommendations. After removing the softmax operation things seem to be working. My model predictions without softmax are around ~5, and deep_lift_shap consistently returns convergence deltas around ~1e-6, which seems pretty good. Interestingly enough when I keep the softmax operation but define it first in the  __init__, deep_lift_shap returns lower convergence deltas than what I had initially (~0.05 for a prediction of ~1 versus my original ~0.8 for a prediction of ~1), but still not as good as just removing the softmax entirely. So it seems the issue was some combination of how the operation was defined and the operation itself (as @jmschrei pointed out), with the latter accounting for the bulk of the discrepancy. Thanks for the help!
